xdsl/dialects/hw.py

```diff
@@ -112,6 +121,142 @@ def print_parameters(self, printer: Printer) -> None:
+
+    def get_inner_symbol_table(self, op: Operation) -> InnerSymbolTable:
+        """Returns the InnerSymolTable trait, ensuring `op` is in the collection"""
+        if not op.has_trait(trait := InnerSymbolTable):
+            raise VerifyException(
+                f"Operation {op.name} should have {trait.__name__} trait"
+            )
+        if op not in self.symbol_tables:
+            self.symbol_tables[op] = InnerSymbolTable(op)
+        return self.symbol_tables[op]
+
+    def populate_and_verify_tables(self, inner_ref_ns_op: Operation):
+        """Populate tables for all InnerSymbolTable operations in the given InnerRefNamespace operation, verifying each."""
+        # Gather top-level operations that have the InnerSymbolTable trait.
+        inner_sym_table_ops = (
+            op for op in inner_ref_ns_op.walk() if op.has_trait(InnerSymbolTable)
+        )
+
+        # Construct the tables
+        for op in inner_sym_table_ops:
+            if op in self.symbol_tables:
+                raise VerifyException(
+                    f"Trying to insert the same op {op} twice in symbol tables"
+                )
+            self.symbol_tables[op] = InnerSymbolTable(op)
+
+
+class InnerRefUserOpInterface(OpTrait):
+    """This interface describes an operation that may use a `InnerRef`. This
+    interface allows for users of inner symbols to hook into verification and
+    other inner symbol related utilities that are either costly or otherwise
+    disallowed within a traditional operation."""
+
+    def verify_inner_refs(self, op: Operation, namespace: "InnerRefNamespace"):
+        """Verify the inner ref uses held by this operation."""
+        ...
+
+
+@dataclass(frozen=True)
+class InnerRefNamespace(OpTrait):
+    """Defines a new scope for InnerRef’s. Operations with this trait myst be a SymbolTable.
+    Combines InnerSymbolTableCollection with a SymbolTable for resolution of InnerRefAttrs, used during verification.
+
+    Inner symbols are more costly than normal symbols, with tricker verification. For this reason,
+    verification is driven as a trait verifier on InnerRefNamespace which constructs and verifies InnerSymbolTables in parallel.
+    See: https://circt.llvm.org/docs/RationaleSymbols/#innerrefnamespace
+    """
+
+    parameters: tuple[SymbolTable, InnerSymbolTableCollection]
+
+    def __init__(
+        self,
+        symbol_table: SymbolTable | None = None,
+        inner_sym_tables: InnerSymbolTableCollection | None = None,
+    ):
+        if symbol_table is None:
+            symbol_table = SymbolTable()
+        if inner_sym_tables is None:
+            inner_sym_tables = InnerSymbolTableCollection()
+        super().__init__((symbol_table, inner_sym_tables))
+
+    @property
+    def symbol_table(self) -> SymbolTable:
+        return self.parameters[0]
+
+    @property
+    def inner_sym_tables(self) -> InnerSymbolTableCollection:
+        return self.parameters[1]
```

Comment:
I think you could just declare `symbol_table` and `inner_sym_tables` at the class level and use dataclass's field(factory=...) feature for the default value. You would get all this code from dataclass.
Or am I missing a reason to use this specific tuple/parameters style?